### ISSUE TITLE
Full TypeScript support for `$`

### ISSUE THREAD

**Issue by luisherranz** [melnikov-s/preact-observables#1]
I have been trying to make it work but have not yet succeeded. It's not clear to me that it's possible.

In principle, going through the object properties and adding a new property with `$` seems possible since TypeScript 4.3:

```ts
export type DeepSignal<T extends object> = {
  [P in keyof T & string as `$${P}`]: T[P] extends object
    ? Signal<DeepSignal<T[P]>>
    : Signal<T[P]>;
} & {
  [P in keyof T]: T[P] extends object ? DeepSignal<T[P]> : T[P];
};
```

But there is a problem:

- If `$` is optional, TypeScript assumes that `store.$prop` may not exist: [playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEkoGdnwMoEsDmA7KEAPACrwC88UuAngHzwDeAUPK4gPa7IAuMArmG7sYACgBuBPiAD8ALnjEAlAG4WbAA4gQAaxGL5xVW3jYQ3eBIhS9Bo22RmLkkOOcGVTAL5MmIAB7qwubc1JrwACJa6lh4BCTw-twguMBo7ABGAFbg3PQUzMYA2gAK8Ji48Nog1OwAZgrwAGTwPDDl2JRoAAYAJD0MxZ5dALpyCiXDCX5JKWlZOWqs0hg4+ESRINGrccQTtLSL8PIxayR7qp5NjIdFpeWV1XUKwwYTUzOp8BnZgvDLG1tYkRdsVhvRXqC7KxPKomGBODx4KBNicCOR4PFEslPt8crQRN93PIAajgXl6N9OhEoqSSLRYfCuOYeFAkuiCmwoPIAIwAGkO6XkHOMrDA8gARAAmcWHbzeOEI5lCODo5GAtYiFlJDxMHjCEAAOh6UGU8AA9Gb4AB3YTaNC1cogXXKw3pI1gU0W622+2O536o1uyxSA1gc2W4DsEDIXAAcnMNpg2h8ergBvS7MQEoAzOL4DDw96k77cCAgA)

  ```ts
  // `$` is optional
  export type DeepSignal<T extends object> = {
    [P in keyof T & string as `$${P}`]?: T[P] extends object
      ? Signal<DeepSignal<T[P]>>
      : Signal<T[P]>;
  } & {
    [P in keyof T]: T[P] extends object ? DeepSignal<T[P]> : T[P];
  };

  const state = {
    a: 1,
    b: {
      c: "2",
    },
  };

  const store = deepSignal(state);

  store.$a; // works fine
  store.b.$c; // works fine
  store.$b.value.c; // doesn't work

  store.b = { c: "3" }; // works fine
  ```

- If `$` is not optional, you cannot mutate the object without adding properties starting with `$`: [playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEkoGdnwMoEsDmA7KEAPACrwC88UuAngHzwDeAUPK4gPa7IAuMArmG7sYACgBuBPiAD8ALnjEAlAG4WbAA4gQAaxGL5xVW3jYQ3eBIhS9Bo22RmLkkOOcGVTAL5MmIAB7qwubc1JrwACJa6lh4BCTw-twguMBo7ABGAFbg3PQUzMYA2gAK8Ji48Nog1OwAZgrwAGTwPDDl2JRoAAYAJD0MxZ5dALoGJcMJfkkpaVk5aqzSGDj4RJEg0StxxOO0tAvw8jGrJLuqnk2MB0Wl5ZXVdQqjCuOT06nwGdmC8Evrm7EiDtisN6GMQXZWJ5VEwwJwePBQBtjgRyPB4olkh8vjlaCIvu55P8UUC8vQvp0IlESSRaDC4VxzDwoEk0QU2FB5ABGAA0B3S8nZxlYYHkACIAExig7ebyw+FMoRwNFIgGrETMpIeJg8YQgAB0PSgyngAHpTfAAO7CbRoWrlEA6pUG9KGsAm81Wm12h1OvWG12WKT6sBmi3WmC2+D23COv1wfXpNmIcUAZjF8GhYcR7BAyFwAHJzBHtEA)

  ```ts
  // `$` is not optional
  export type DeepSignal<T extends object> = {
    [P in keyof T & string as `$${P}`]: T[P] extends object
      ? Signal<DeepSignal<T[P]>>
      : Signal<T[P]>;
  } & {
    [P in keyof T]: T[P] extends object ? DeepSignal<T[P]> : T[P];
  };

  const state = {
    a: 1,
    b: {
      c: "2",
    },
  };

  const store = deepSignal(state);

  store.$a; // works fine
  store.b.$c; // works fine
  store.$b.value.c; // works fine

  store.b = { c: "3" }; // doesn't work
  ```

Are you familiar with this problem? Any ideas on how to fix it?

By the way, I tested your library and it looks like you are just adding `$` to the getters, aren't you? Is it for some reason? What is your idea regarding TypeScript?

_I've recreated the example you have in the README in [Stackblitz](https://stackblitz.com/edit/vitejs-vite-nbhvq5?file=src%2Fmain.tsx&terminal=dev), and added a few `$` to test this. I'm not sure if I'm doing it right._

**Comment by luisherranz** [comment]
I wonder if we need to wait for [something like this](https://github.com/microsoft/TypeScript/issues/43826), or if there is a workaround that we can use today.

**Comment by luisherranz** [comment]
By the way, the two APIs that could work are (as far as I know):

- Use a function to get the signals in TS (not required in JS)

  ```js
  store.$b.value.c; // doesn't in TS, but works in JS
  signal(store, "b").value.c; // works fine in both
  ```

  [You've tested that](https://github.com/melnikov-s/preact-observables/blob/main/src/index.ts#L120), haven't you?

- Use a setter to mutate the objects in TS (not required in JS): [playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEkoGdnwMoEsDmA7KEAPACrwC88UuAngHzwDeAUPK4gPa7IAuMArmG7sYACgBuBPiAD8ALnjEAlAG4WbAA4gQAaxGL5xVW3jYQ3eBIhS9Bo22RmLkkOOcGVTAL5MmIAB7qwubc1JrwACJa6lh4BCTw-twguMBo7ABGAFbg3PQUzMYA2gAK8Ji48Nog1OwAZgrwAGTwPDDl2JRoAAYAJD0MxZ5dALoGJcMJfkkpaVk5aqzSGDj4RJEg0StxxOO0tAvw8jGrJLuqnk2MB0Wl5ZXVdQqjCuOT06nwGdmC8Evrm7EiDtisN6GMQXZWJ5VEwwJwePBQBtjgRyPB4olkh8vjlaCIvu55P8UUC8vQvp0IlESSRaKo4VxzA5zBQMVMsbNvrl8VlCVTkVtSeRyVlKcTBbSYQyETwoEk0QU2FB5ABGAA0B3S8kVxlYYHkACIAEwGg7ebyw+FMoRwNFIgGrESypIeJg8YQgAB0PSgyngAHp-fAAO7CbRoWrlEBum1e9LesB+wMhsMRqMxj3e+OWKSesABoOhmDh+CR3DRjNwT3pBWIQ0AZgN8GhBcR7BAyFwAHJzEXtJW42jmSIGHX4AbG83FK2+2ny0A)

  ```ts
  const set = <T extends object>(obj: T): DeepSignal<T> => obj as DeepSignal<T>;

  store.b = { c: "3" }; // doesn't work in TS, but it works in JS
  store.b = set({ c: "3" }); // works fine in both
  ```

Of course, I'd prefer to avoid them 🙂

**Comment by melnikov-s** [comment]
Thanks for creating this issue! 

It completely slipped my mind that TS will block you on assignment as it now believes that `$whatever` is part of the type and therefore should be present. I also did not deeply apply the type so only the top level properties had `$` on them, oops! 

I think you're right that there's no way to address this currently in TS. Either we go for an optional `$` type which will require the consumer to assert that it is not null/undefined (`!`) or we are stuck with needing special setters to perform mutations. 

I've opted to make the `$` properties optional, I feel like that's the lesser of the two evils here especially since I figure the vast majority of the time consumers will use `$` to pass signals to text nodes and dom attributes and react/preact will happy accept `undefined` there, so no need for a `!` assertion.

I've published a new version (0.0.5) with these changes

**Comment by luisherranz** [comment]
> the vast majority of the time consumers will use $ to pass signals to text nodes and dom attributes and react/preact will happy accept undefined there

That makes sense.

What do you think, @darerodz?

**Comment by luisherranz** [comment]
By the way, I updated the [Stackblitz](https://stackblitz.com/edit/vitejs-vite-nbhvq5?file=src%2Fmain.tsx&terminal=dev) to the latest version and it works great now 🙂

**Comment by DAreRodz** [comment]
Hey @luisherranz, thanks for pinging me (although I'm pretty sure you and @melnikov-s have more context on this issue than me, haha 😅). Well, I can give my two cents.

If we have to choose either an optional `$` for signals or a special setter to perform mutations, I will select the first option. For me, the main idea of a library like this would be to allow using an object ―based on signals― of any shape, as transparently as possible, as if it were a regular object. Having to use a setter wrapper in TS to modify the object, which is expected to happen frequently, makes me think such an API may be a bit cumbersome.

I also agree with @melnikov-s when he says that, most of the time, you pass signals to text nodes and dom attributes and that there are no issues with signals being typed as `undefined`. There could be problems, though, if you need them to be correctly typed (e.g., if you want to use `peek()`). I have to admit that I don't know how usual would be such cases, but I'm fine using `!` for those. 🙂

**Comment by luisherranz** [comment]
Great. Thank you, folks. I think we'll also go with that option.

@melnikov-s, I'm going to close this issue but if you find more TS problems with this solution in the future please let us know 🙂

**Comment by luisherranz** [comment]
@melnikov-s, I've finally published the package: https://github.com/luisherranz/deepsignal/

I'm happy with how TypeScript ended up working. Thanks for your help. 

I ended up using `array.$[0]` instead of `array.$0` and `peek(store, "prop")` to peek, but other than that, it follows what we discussed here.

I haven't tested the library in a real project yet tho, so there may still be issues.

This is a [Stackblitz example](https://stackblitz.com/edit/vitejs-vite-hktyyf?file=src%2Fmain.tsx) using TypeScript.
